### ISSUE TITLE
AGS4: Eliminate special Inventory MouseButton values 

### ISSUE THREAD

**Issue by messengerbag** [adventuregamestudio/ags#1283]
**Describe the problem**
Currently (as of 3.5), the `MouseButton` enum features special values for when the mouse is clicked over an inventory window: `eMouseLeftInv`, `eMouseMiddleInv`, and `eMouseRightInv`.

This is a pointless and unintuitive complication (presumably a legacy feature from way back in the day to solve some issue that is no longer relevant). It makes it harder to write UI code, and it trips up newbies who aren't aware of it.

**Suggested change**
Remove these values, and change the behavior so that `Mouse.IsButtonDown()`and `on_mouse_click()` report button presses consistently regardless of what the cursor is over.

As part of this change, you may also want to consider getting rid of "built-in inventory window click handling" altogether, and therefore also the project option "Override built-in inventory window click handling."

**Comment by ivan-mogilko** [comment]
There was another related ticket here: #472, these may be considered together.

**Comment by ivan-mogilko** [comment]
I am considering doing this in AGS 4 branch now, but I'd like to first discuss how on_mouse_click would behave, what are possible complications for users, and whether we have everything necessary to implement wanted behavior in script right now.

As everyone should now, `on_mouse_click` is not called at all when clicking on GUI, it's called only when the click went to the room. Every existing game script accounts for that.

Suppose we have mouse events that are called always, regardless of where the player clicked. How would a user deal with that in script? They will have to separate gui clicks from room clicks. I can see two existing methods that allow to do so: `GetLocationType` that returns None for GUI, and checking result of GUI.GetAtScreenXY.

After this, if user is to code clicking on inventory, there's three methods: a legacy `game.inv_activated` variable, that stores the index of a last item clicked upon, InventoryItem.GetAtScreenXY, and then also InventoryWindow.GetAtScreenXY (going this route requires to calculate item under mouse, so there has to be a good reason to use this method).

Questions so far:
1. Do we need to add anything else, or existing methods are enough?
2. Do we need to keep a room-only click callback, maybe add gui-only click callback, for convenience?

EXTRA NOTE: I am actually planning on adding separate mouse_down and mouse_up script callbacks, but not sure if it's okay to keep mouse_click instead mouse_down, as it's already called when the mouse is down (only when above rooms). So this problem also related to the above questions.

**Comment by ericoporto** [comment]
Hey, a few notes

> As everyone should now, on_mouse_click is not called at all when clicking on GUI

Actually, for an inventory, it is called, which can be differentiated by the different mouse buttons today - my guess is orignal events in the engine only had one parameter. Today, I think the minimal global script is probably something like 

```AGS Script
function on_mouse_click(MouseButton button)
{
  if (button == eMouseLeftInv || button == eMouseRightInv || button == eMouseMiddleInv)
  {
    handle_inventory_click(button);
  }
  else if (button == eMouseLeft)
  {
    // left-click, so try using the current mouse cursor mode at this position
    Room.ProcessClick(mouse.x, mouse.y, mouse.Mode);
  }
}
```

I think GUI clicks are handled magically?

> EXTRA NOTE: I am actually planning on adding separate mouse_down and mouse_up script callback

Just to check, because I forgot, are the on_pointer_down, on_pointer_up, on_pointer_move already triggering when hitting anywhere in the screen? And I don't remember how they work if we have the touch emulating a mouse instead of being a touch.

There is a forum thread (and a few others) of people finding this quirk: https://www.adventuregamestudio.co.uk/forums/beginners-technical-questions/working-as-designed-on_mouse_click-not-called-when-clicking-on-gui/msg636457773/#msg636457773

This one above reminded me of the `on_event` with `eEventGUIMouseDown` and `eEventGUIMouseUp`.

On mouse click quirks, the other one is that `AnyClick` interaction on the things in the room is actually `AnyClickWhenNotWalkMode` - and a lot of games are coding assuming this, but it does look a bit mistic if you forget about it. I think this should be kept though because people are really used to it, and updating the games here would requiring fixing code that is in many rooms...

These quirks though are known by people using AGS for ages, so I am not sure on how much to change on these things without breaking people's code. But wanted to note a few things here just to note them somewhere... You also make an interesting alternative here: https://github.com/adventuregamestudio/ags/issues/472#issuecomment-412046036

**Comment by edmundito** [comment]
My first thought is that it should be an event handler for the InvWindow control. As for backwards compatibility, if any, it should be a toggle in the settings like any other breaking change.

**Comment by ivan-mogilko** [comment]
@ericoporto 
> Actually, for an inventory, it is called, which can be differentiated by the different mouse buttons today

Yes, but this issue is exactly for removing these "mouse buttons" (the current behavior is already explained in the first post), so I skipped them.

> On mouse click quirks, the other one is that AnyClick interaction on the things in the room is actually AnyClickWhenNotWalkMode - and a lot of games are coding assuming this, but it does look a bit mistic if you forget about it. I think this should be kept though because people are really used to it, and updating the games here would requiring fixing code that is in many rooms...

This is related rather to how Cursor modes work, and was addressed by removing hardcoded cursor modes earlier (#2860).

**Comment by ivan-mogilko** [comment]
> My first thought is that it should be an event handler for the InvWindow control.

Right, i forgot about this option; this may be also implemented. But I am more interested how would a generic handling look like.
